src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,275 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+#if CORE_PCL || NET45 || NET46
+    using TaskEx = System.Threading.Tasks.Task;
+#endif
+
+    /// <summary>
+    /// Represents a hub for metric aggregation procedures.
+    /// </summary>
+    public sealed class MetricAggregatorManager : IDisposable
+    {
+        /// <summary>
+        /// Name of the property added to aggregation results to indicate duration of the aggregation interval.
+        /// </summary>
+        private static string intervalDurationPropertyName = "IntervalDurationMs";
+
+        /// <summary>
+        /// Reporting frequency.
+        /// </summary>
+        private static TimeSpan snapshotFrequency = TimeSpan.FromMinutes(1);
+
+        /// <summary>
+        /// Last time snapshot was initiated.
+        /// </summary>
+        private DateTimeOffset lastSnapshotStartDateTime;
+
+        /// <summary>
+        /// Cancellation token source to allow cancellation of the snapshotting task.
+        /// </summary>
+        private CancellationTokenSource cancellationSource;
+
+        /// <summary>
+        /// Metric aggregator snapshotting task.
+        /// </summary>
+        private Task snapshotTask;
```

Comment:
I looked at TaskTimer. I understand what it is doing and it would [almost] work for me if I were to have functionality (in it) to do something if wait is terminated (I need to try to do Flush() to attempt not to lose any telemetry I can salvage). I considered the problem and I think it is too large to be coupled with the change I'm doing.

I suggest separating the area that covers two questions: A) How many timer implementations does one library need? and B) How many utility threads does one library need to run? and tackling it separately. We also have a perfect guy to propose the best solution - Greg. I chatted with him and once he is done with his current thing (should be just a few days) he'll look into this.